src/runner/systemdconn.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <chrono>
+#include <systemd/sd-bus-protocol.h>
+#include <thread>
+
+#include <aos/common/tools/memory.hpp>
+#include <logger/logmodule.hpp>
+#include <utils/exception.hpp>
+
+#include "systemdconn.hpp"
+
+namespace aos::sm::runner {
+
+/***********************************************************************************************************************
+ * Statics
+ **********************************************************************************************************************/
+
+namespace {
+
+RetWithError<InstanceRunState> ConvertToInstanceState(const std::string& src)
+{
+    if (src == "active") {
+        return {InstanceRunStateEnum::eActive};
+    }
+
+    // Treat all other statuses as failed: reloading, inactive, failed, activating, deactivating.
+    return {InstanceRunStateEnum::eFailed};
+}
+
+aos::Duration ToUSec(const Duration& val)
+{
+    return val / 1000;
+}
+
+} // namespace
+
+/***********************************************************************************************************************
+ * Implementation
+ **********************************************************************************************************************/
+
+SystemdConn::SystemdConn()
+{
+    // According to https://www.freedesktop.org/software/systemd/man/255/sd_bus_open.html
+    // sd_bus_open_system creates connection object associated with the calling thread,
+    // Consequently it might be unsafe to use it from multiple threads even with proper synchronization.
+    // Still leave the current approach as no problems observed yet.
+    auto rv = sd_bus_open_system(&mBus);
+    if (rv < 0) {
+        AOS_ERROR_THROW(strerror(-rv), Error(-rv));
+    }
+}
+
+SystemdConn::~SystemdConn()
+{
+    sd_bus_unref(mBus);
+}
+
+RetWithError<std::vector<UnitStatus>> SystemdConn::ListUnits()
+{
+    std::unique_lock lock {mMutex};
+
+    sd_bus_error    error   = SD_BUS_ERROR_NULL;
+    sd_bus_message* reply   = nullptr;
+    auto            freeErr = DeferRelease(&error, sd_bus_error_free);
+    auto            freeMsg = DeferRelease(reply, sd_bus_message_unref);
+
+    auto rv = sd_bus_call_method(mBus, cDestination, cPath, cInterface, "ListUnits", &error, &reply, nullptr);
+    if (rv < 0) {
+        LOG_ERR() << "err: " << error.name << ", msg: " << error.message;
```

Comment:
Add meaningful message instead of err. Or better return error with message:

`return {{}, Error(rv, "your err message")};`